pip/wheel.py

```diff
@@ -631,26 +657,126 @@ def supported(self, tags=None):
         return bool(set(tags).intersection(self.file_tags))
 
 
+class BuildEnvironment(object):
+    """Context manager to install build deps in a simple temporary environment
+    """
+    def __init__(self, no_clean=False):
+        self.prefix = tempfile.mkdtemp('pip-build-env-')
+        self.no_clean = no_clean
+
+    def __enter__(self):
+        self.save_path = os.environ.get('PATH', None)
+        self.save_pythonpath = os.environ.get('PYTHONPATH', None)
+
+        install_scheme = 'nt' if (os.name == 'nt') else 'posix_prefix'
+        install_dirs = get_paths(install_scheme, vars={
+            'base': self.prefix,
+            'platbase': self.prefix,
+        })
+
+        scripts = install_dirs['scripts']
+        if self.save_path:
+            os.environ['PATH'] = scripts + os.pathsep + self.save_path
+        else:
+            os.environ['PATH'] = scripts + os.pathsep + os.defpath
+
+        if install_dirs['purelib'] == install_dirs['platlib']:
+            lib_dirs = install_dirs['purelib']
+        else:
+            lib_dirs = install_dirs['purelib'] + os.pathsep + \
+                install_dirs['platlib']
+        if self.save_pythonpath:
+            os.environ['PYTHONPATH'] = lib_dirs + os.pathsep + \
+                self.save_pythonpath
+        else:
+            os.environ['PYTHONPATH'] = lib_dirs
+
+        return self.prefix
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.save_path is None:
+            os.environ.pop('PATH', None)
+        else:
+            os.environ['PATH'] = self.save_path
+
+        if self.save_pythonpath is None:
+            os.environ.pop('PYTHONPATH', None)
+        else:
+            os.environ['PYTHONPATH'] = self.save_pythonpath
+
+        if not self.no_clean:
+            rmtree(self.prefix)
+
+        return False  # Do not suppress exceptions
+
+
 class WheelBuilder(object):
     """Build wheels from a RequirementSet."""
 
     def __init__(self, requirement_set, finder, build_options=None,
-                 global_options=None):
+                 global_options=None, no_clean=False):
         self.requirement_set = requirement_set
         self.finder = finder
         self._cache_root = requirement_set._wheel_cache._cache_dir
         self._wheel_dir = requirement_set.wheel_download_dir
         self.build_options = build_options or []
         self.global_options = global_options or []
+        self.no_clean = no_clean
+
+    def _find_build_reqs(self, req):
+        """Get a list of the packages required to build the project, if any.
+
+        Build requirements can be specified in a pyproject.toml, as described
+        in PEP 518
+        """
+        if os.path.isfile(req.pyproject_toml):
+            with open(req.pyproject_toml) as f:
+                pp_toml = pytoml.load(f)
+            return pp_toml.get('build-system', {}).get('requires', [])
+
+        return []  # No pyproject.toml
```

Comment:
Yeah... Though I'm actually flip flopping a bit mentally. The spec actually says that `build-system.requires` is mandatory in all pyproject.toml files. It's annoying if people are already creating invalid files; are you sure we can't break them? 😛

The discussion below about isolation reminds us too that we need some trigger for enabling isolated builds. I think the original idea was that either you would have a pyproject.toml and it would have explicit requires and you would get an isolated build, or else you would have none of those. If we have to treat pyproject.toml's that are missing the build-system section as legacy packages, then we need something different.

Of course even if we were able to make build-system.requires mandatory, we'd still have to default on the actual build system key (the one that's not specified yet). This is a late and somewhat wacky idea, but I guess we could say that it's mandatory to have a key `build-system = "setuptools"`, exactly that string, and then later we could redefine it as the package that gets imported to provide the build system API once that's defined...